### ISSUE TITLE
Terraform Crash using data aws_subnet_ids

### ISSUE THREAD

**Issue by jimbrockman** [hashicorp/terraform#18240]
### Terraform Version

```
Terraform v0.11.7
+ provider.aws v1.22.0
```

### Terraform Configuration Files

variables.tf
```hcl
ariable "created_by" {}
variable "aws_access_key" {}
variable "aws_secret_key" {}

variable "query_aws" {
  default = true
}

data "aws_vpc" "admcsa_vpc" {
  count = "${var.query_aws}"

  filter {
    name   = "tag:AppOwner"
    values = ["ADMCSA"]
  }
}

data "aws_subnet_ids" "admcsa_private_subnet_ids" {
  count  = "${var.query_aws}"
  vpc_id = "${data.aws_vpc.admcsa_vpc.*.id}"

  tags {
    AppOwner = "ADMCSA"
  }
}

variable "dev_hardcoded_vpc" {
  default = ["vpc-123456"]
}

variable "dev_hardcoded_subnet" {
  default = ["subnet-123456"]
}

#output "vpc_id" {
#  value = ["${var.query_aws == true ? data.aws_vpc.admcsa_vpc.*.id : var.dev_hardcoded_vpc}"]
# }
output "vpc_id" {
  # If we're querying AWS to get VPC ID, return that.  Otherwise return configured vpc-id
  value = "${element(coalescelist(data.aws_vpc.admcsa_vpc.*.id, var.dev_hardcoded_vpc), 1)}"
}

output "private_subnet_ids" {
  value = "${element(coalescelist(data.aws_subnet_ids.admcsa_private_subnet_ids.private.ids, var.dev_hardcoded_subnet), 1)}"
}
```
main.tf
```
provider "aws" {
  region     = "us-east-1"
  access_key = "${var.aws_access_key}"
  secret_key = "${var.aws_secret_key}"
}
```

### Crash Output

https://gist.github.com/jimbrockman/545cae71cf2c7099c77b39cad5ee44ce

### Actual Behavior

```
data.aws_vpc.admcsa_vpc: Refreshing state...

Error: Error refreshing state: 1 error(s) occurred:

* data.aws_subnet_ids.admcsa_private_subnet_ids: 1 error(s) occurred:

* data.aws_subnet_ids.admcsa_private_subnet_ids: unexpected EOF


panic: Error reading level config: '' expected type 'string', got unconvertible type '[]interface {}'
```

### Steps to Reproduce

1. `terraform init`
2. `terraform workspace select dev-foo`
3. `terraform apply`

**Comment by apparentlymart** [comment]
Hi @jimbrockman! Sorry for this crash, and thanks for reporting it.

It looks like the trigger for this crash is that you've assigned a list to the `vpc_id` attribute, which is expecting a string. This shouldn't be a crash, of course: Terraform should've caught this during validation.

For now, I think you should be able to work around this by making sure to assign a single string:

```hcl
data "aws_subnet_ids" "admcsa_private_subnet_ids" {
  count  = "${var.query_aws}"
  vpc_id = "${data.aws_vpc.admcsa_vpc.*.id[count.index]}"

  tags {
    AppOwner = "ADMCSA"
  }
}
```

The validation code has been significantly changed in our development branch for the next major release of Terraform, and so I suspect that this missing validation has already been fixed. We'll use this issue as a reminder to test once the branch is merged to verify that it is indeed fixed.

Thanks again for reporting this!

**Comment by jimbrockman** [comment]
Thanks!  I also found that removing the splat syntax also works: 

```data "aws_subnet_ids" "admcsa_private_subnet_ids" {
  count  = "${var.query_aws}"
  vpc_id = "${data.aws_vpc.admcsa_vpc.id}"

  tags {
    AppOwner = "ADMCSA"
  }
}
```

**Comment by apparentlymart** [comment]
Great! Please note that the example you've posted in your latest comment will use the _first_ `aws_vpc` result for all of the `aws_subnet_ids`, which is different than my example which would correlate the VPCs to the subnets using their count indexes.

In other words, writing `data.aws_vpc.admcsa_vpc.id` is, in this case, a synonym for `data.aws_vpc.admcsa_vpc.*.id[0]`, rather than `data.aws_vpc.admcsa_vpc.*.id[count.index]`.

**Comment by apparentlymart** [comment]
Hi @jimbrockman!

Having now looked through historical issues, I can see that this was covering the same root cause as #16902, which is now resolved with a fix that will be included in the forthcoming v0.12.0 release. You can see in my comment in that issue how I verified it with the v0.12.0-alpha1 prerelease build.

Since there's a fix in for this, I'm going to close this out now. Thanks again for reporting this, and for your patience while we got to the bottom of it.
